### PR TITLE
[improvement](mtmv) Support to partition prune when query rewrite by sync materialized view

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Support to partition prune when query rewrite by sync materialized view
such as table def is as following:
```sql
        CREATE TABLE IF NOT EXISTS test_unique (
        `time` DATETIME NULL COMMENT '查询时间', 
        `app_name` VARCHAR(64) NULL COMMENT '标识', 
        `event_id` VARCHAR(128) NULL COMMENT '标识', 
        `decision` VARCHAR(32) NULL COMMENT '枚举值', 
        `id` VARCHAR(35) NOT NULL COMMENT 'od', 
        `code` VARCHAR(64) NULL COMMENT '标识', 
        `event_type` VARCHAR(32) NULL COMMENT '事件类型' 
        )
        UNIQUE KEY(time)
        PARTITION BY RANGE(time)                                                                                                                                                                                                                
        (                                                                                                                                                                                                                                      
         FROM ("2024-07-01 00:00:00") TO ("2024-07-15 00:00:00") INTERVAL 1 HOUR                                                                                                                                                                              
        )     
        DISTRIBUTED BY HASH(time)
        BUCKETS 3 PROPERTIES ("replication_num" = "1");
```
sync materialized view def is

```sql
create materialized view as
    select
    app_name,
    event_id,
    time,
    count(*)
    from 
    test_duplicate
    group by
    app_name,
    event_id,
    time;
```

if your query is following, if rewritten by sync materialized view successfully, should partition prune
```sql
    select
    app_name,
    event_id,
    time,
    count(*)
    from 
    test_duplicate
    where time < '2024-07-05 01:00:00'
    group by
    app_name,
    time,
    event_id;

```

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/SyncMaterializationContext.java [hunk 1]
```diff
@@ -96,13 +99,29 @@ String getStringInfo() {
...
         return Optional.of(Pair.of(relationId, normalizeStatisticsColumnExpression(statistics)));
     }
 
+    @Override
+    public Plan getScanPlan(StructInfo queryStructInfo) {
+        if (queryStructInfo == null) {
+            return scanPlan;
+        }
+        if (queryStructInfo.getRelations().size() == 1
+                && queryStructInfo.getRelations().get(0) instanceof LogicalOlapScan
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/SyncMaterializationContext.java:116]: refer to `org.apache.doris.nereids.rules.rewrite.DeferMaterializeTopNResult`, maybe relation could be `LogicalDeferMaterializeOlapScan`? can we collect correct struct info?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializedViewUtils.java [hunk 2]
```diff
@@ -224,35 +223,20 @@ public static List<StructInfo> extractStructInfo(Plan plan, Plan originalPlan, C
...
-                cascadesContext.getStatementContext().getNextRelationId(),
-                olapTable,
-                ImmutableList.of(olapTable.getQualifiedDbName()),
-                // this must be empty, or it will be used to sample
-                ImmutableList.of(),
-                olapTable.getPartitionIds(),
+                partitionIds,
                 indexId,
-                PreAggStatus.unset(),
+                preAggStatus,
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializedViewUtils.java:237]: this is preAggStatus is variable. 
it's is unset when sync materialization context and is on when async materialization context

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/SyncMaterializationContext.java (hunk 1 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/SyncMaterializationContext.java:116]: Yeah, we can collect the `LogicalDeferMaterializeOlapScan` when collect the struct info
```java
private static class RelationCollector extends DefaultPlanVisitor<Void, List<CatalogRelation>> {
        @Override
        public Void visit(Plan plan, List<CatalogRelation> collectedRelations) {
            if (plan instanceof CatalogRelation) {
                collectedRelations.add((CatalogRelation) plan);
            }
            return super.visit(plan, collectedRelations);
        }
    }
```

**Comment by seawinde** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]

**Review by zddr — APPROVED** [review]

**Pull request merged**